cpu-sim/src/lib.rs

```diff
@@ -1,12 +1,62 @@
+pub mod bus;
+pub mod dram;
+pub mod fifo;
 pub mod memory;
 pub mod sim;
 
 pub use sim::SimulationResult;
 
-use memory::Memory;
+use bus::SystemBus;
+use dram::Dram;
 use sim::Simulator;
 use std::path::Path;
 
+/// Run an ELF file on the simulated CPU with an optional FIFO callback
+///
+/// # Arguments
+/// * `elf_path` - Path to the RISC-V ELF executable
+/// * `max_cycles` - Maximum number of cycles to run
+/// * `print_inst_trace` - Whether to print instruction trace
+/// * `fifo_callback` - Optional callback invoked when data is written to the FIFO
+///
+/// # Returns
+/// * `Ok(SimulationResult)` on success
+/// * `Err(String)` on error
+pub fn run_elf_with_callback<F>(
+    elf_path: &Path,
+    max_cycles: u64,
+    print_inst_trace: bool,
+    fifo_callback: Option<F>,
+) -> Result<SimulationResult, String>
+where
+    F: FnMut(u8),
```

Comment:
This callback needs to be expanded to a u32 since the fifo only works with full 32 bit word values.